src/ds3.c

```diff
@@ -121,6 +125,140 @@ void ds3_client_register_logging(ds3_client* client, ds3_log_lvl log_lvl, void (
     client->log = log;
 }
 
+static void _ds3_free_metadata_entry(gpointer pointer) {
+    ds3_metadata_entry* entry;
+    if (pointer == NULL) {
+        return; // do nothing
+    }
+
+    entry = (ds3_metadata_entry*) pointer;
+
+    ds3_free_metadata_entry(entry);
+}
+
+/*
+ * This copies all the header values in the ds3_response_header struct so that they may be safely returned to the user
+ * without having to worry about if the data is freed internally.
+ */
+static ds3_metadata_entry* ds3_metadata_entry_init(ds3_response_header* header) {
+    uint i;
+    ds3_str* header_value;
+    GPtrArray* values = g_ptr_array_new();
+    ds3_str* key_name;
+    ds3_metadata_entry* response = g_new0(ds3_metadata_entry, 1);
+
+    for (i = 0; i < header->values->len; i++) {
+        header_value = g_ptr_array_index(header->values, i);
+        g_ptr_array_add(values, ds3_str_dup(header_value));
+    }
+
+    key_name = ds3_str_init(header->key->value + 11);
+
+    response->num_values = values->len;
+    response->name = key_name;
+    response->values = (ds3_str**) g_ptr_array_free(values, FALSE);
+    fprintf(stderr, "creating metadata entry of: %s\n", key_name->value);
+    return response;
+}
+
+/* The headers hash table contains all the response headers which have the following types:
+ * Key - char*
+ * Value - ds3_response_header
+ *
+ * All values should be copied from the struct to avoid memory issues
+ */
+static ds3_metadata* _init_metadata(GHashTable* response_headers) {
+    struct _ds3_metadata* metadata = g_new0(struct _ds3_metadata, 1);
+    GHashTableIter iter;
+    gpointer _key, _value;
+    char* key;
+    ds3_response_header* header;
+    ds3_metadata_entry* entry;
+    metadata->metadata = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, _ds3_free_metadata_entry);
+
+    if (response_headers == NULL) {
+        fprintf(stderr, "response headers was null\n");
+    }
+
+    g_hash_table_iter_init(&iter, response_headers);
+
+    while(g_hash_table_iter_next(&iter, &_key, &_value)) {
+        key = (char*) _key;
+        header = (ds3_response_header*) _value;
+        if (g_str_has_prefix(key, "x-amz-meta-")) {
+            entry = ds3_metadata_entry_init(header);
+            g_hash_table_insert(metadata->metadata, g_strdup(entry->name->value), entry);
+        }
+    }
+
+    return (ds3_metadata*) metadata;
+}
+
+ds3_metadata_entry* ds3_metadata_get_entry(const ds3_metadata* _metadata, const char* name) {
+    ds3_metadata_entry* copy;
+    ds3_metadata_entry* orig;
+    ds3_str** metadata_copy;
+    uint64_t i;
+    struct _ds3_metadata* metadata = (struct _ds3_metadata*) _metadata;
```

Comment:
As a note to others, since Denver and I talked in person.  The reason for this is to be consistent with other places that we do this, and to be very explicit as to what the type actually is.